### PR TITLE
Explicit core number in topology components

### PR DESCRIPTION
**Pull request by lyakh**
Make all components respect the DSP core ID, indicated in the `PIPELINE_PCM_ADD()` macro. This will create a series of patches, that will:
1. add a `SOF_TKN_COMP_CORE_ID` token to be used in all components
2. make all the components use the core ID from `PIPELINE_PCM_ADD()` (and possibly from other additional sources, e.g. from `DAI_ADD()`) with that token

After that the next step has to be decided upon. The original plan was to:

3. switch the "scheduler" widget to also use `SOF_TKN_COMP_CORE_ID` instead of the currently used `SOF_TKN_SCHED_CORE`
4. deprecate and remove `SOF_TKN_SCHED_CORE`

But that might not be needed. After all that we'll move on to adding a possibility to place single components on specific cores and add support for that in the kernel.

### CODE REVIEW

**Comment by plbossart** [comment]
> 3\. switch the "scheduler" widget to also use `SOF_TKN_COMP_CORE_ID` instead of the currently used `SOF_TKN_SCHED_CORE`
> 4\. deprecate and remove `SOF_TKN_SCHED_CORE`

What exactly would be the difference between SOF_TKN_COMP_CORE_ID and SOF_TKN_SCHED_CORE?

I am concerned about support for old topologies with new kernels. Adding/removing tokens is problematic, e.g. we screwed-up royally with the HDaudio/LED tokens so let's be careful, shall we?

File: tools/topology/m4/bytecontrol.m4 [hunk 1]
```diff
@@ -32,7 +32,7 @@ define(`CONTROLBYTES_PRIV',
 `	$2'
 `}')
 
-dnl C_CONTROLMIXER(name, index, ops, base, num_regs, mask, max, tlv, priv)
+dnl C_CONTROLBYTES(name, index, ops, base, num_regs, mask, max, tlv, priv)
```
Comment (ranj063) [inline tools/topology/m4/bytecontrol.m4]: @lyakh all the above should be in a separate patch?

File: tools/topology/m4/mixercontrol.m4 [hunk 2]
```diff
@@ -31,20 +31,30 @@ define(`CONTROLMIXER_OPS',
...
 `	}'
 `}'
 `SectionData."'N_CONTROLMIXER($2)`_data_w" {'
 `	tuples "'N_CONTROLMIXER($2)`_tuples_w"'
 `}'
 ,` ')'
+`SectionVendorTuples."'N_CONTROLMIXER($2)`_tuples_comp_w" {'
+`	tokens "sof_comp_tokens"'
+`	tuples."word" {'
+`		SOF_TKN_COMP_CORE_ID'		STR($9)
```
Comment (ranj063) [inline tools/topology/m4/mixercontrol.m4]: is this correct? The comment shows the core is the last argument?

**Comment by lgirdwood** [comment]
> > 3. switch the "scheduler" widget to also use `SOF_TKN_COMP_CORE_ID` instead of the currently used `SOF_TKN_SCHED_CORE`
> > 4. deprecate and remove `SOF_TKN_SCHED_CORE`
> 
> What exactly would be the difference between SOF_TKN_COMP_CORE_ID and SOF_TKN_SCHED_CORE?
> 

We need a way to differentiate between scheduling a whole pipeline on core X and a single component on core X. We could alias some of the current token names to include the differentiation, i.e.. add a SOF_TKN_PIPELINE SCHED_CORE to alias SOF_TKN_SCHED_CORE.

> I am concerned about support for old topologies with new kernels. Adding/removing tokens is problematic, e.g. we screwed-up royally with the HDaudio/LED tokens so let's be careful, shall we?

Agree, needs to be addressed in IPC2.0

**Comment by juimonen** [comment]
mute led issues we're caused by new topologies with old kernels, but yeah it's bad and we really should avoid that.

File: tools/topology/m4/mixercontrol.m4 (hunk 2 above)
Comment (lyakh) [reply tools/topology/m4/mixercontrol.m4]: @ranj063 the comment is wrong, thanks!

File: src/schedule/edf_schedule.c [hunk 3]
```diff
@@ -124,12 +129,16 @@ static void schedule_edf_task(void *data, struct task *task, uint64_t start,
...
 	/* add task to the list */
 	list_item_append(&task->list, &edf_sch->list);
+	tracev_edf_sch("schedule_edf_task() %p", (long)task);
 
 	task->state = SOF_TASK_STATE_QUEUED;
 
 	irq_local_enable(flags);
 
-	schedule_edf(data);
+	if (cpu_get_id() == PLATFORM_MASTER_CORE_ID ||
```
Comment (tlauda) [inline src/schedule/edf_schedule.c]: @lyakh Nice catch. It's because master core doesn't wait for slave core to actually boot up before responding to power up ipc. I wonder whether more appropriate solution would be to wait for some kind of status reported by slave core after starting. This feels a little bit hacky.

File: tools/topology/m4/pga.m4 [hunk 4]
```diff
@@ -6,13 +6,14 @@ dnl PGA name)
 
 define(`N_PGA', `PGA'PIPELINE_ID`.'$1)
 
-dnl W_PGA(name, format, periods_sink, periods_source, kcontrol0. kcontrol1...etc)
+dnl W_PGA(name, format, periods_sink, periods_source, core, kcontrol0. kcontrol1...etc)
```
Comment (lgirdwood) [inline tools/topology/m4/pga.m4]: Btw, unrelated (and not for this PR) - can you start thinking about how we can reduce the number of arguments for each macro. I'm thinking given the W_PGA example above we could have
```
W_PERIODS(name, PGA, 2,2)
W_FORMAT(name, PGA, s32le)
W_KCONTROLS(name, PGA, ...)
```
This would then define macros 
```
define(`PGA_name_PERIODS_SOURCE`, `2`)
```
and so on.
These could all then be brought together and instantiated using
```
W_PGA(name)
```
Which then uses all the PGA_name_ macros to construct the PGA. This also means if we dont define a W_CORE() macro then the default value is used or it's left blank. This means it becomes much easier to expand the functionality and simplifies usage.

File: src/audio/dai.c [hunk 5]
```diff
@@ -203,9 +203,9 @@ static void dai_free(struct comp_dev *dev)
 	rfree(dev);
 }
 
-static inline int dai_comp_get_hw_params(struct comp_dev *dev,
-					 struct sof_ipc_stream_params *params,
-					 int dir)
+static int dai_comp_get_hw_params(struct comp_dev *dev,
```
Comment (lgirdwood) [inline src/audio/dai.c]: Best make this separate PR.

File: src/ipc/handler.c [hunk 6]
```diff
@@ -1044,7 +1044,12 @@ static int ipc_glb_tplg_comp_new(uint32_t header)
 {
 	struct ipc *ipc = ipc_get();
 	struct sof_ipc_comp comp;
-	struct sof_ipc_comp_reply reply;
+	struct sof_ipc_comp_reply reply = {
```
Comment (lgirdwood) [inline src/ipc/handler.c]: Another separate PR.

File: src/ipc/ipc.c [hunk 7]
```diff
@@ -211,7 +210,7 @@ int ipc_comp_new(struct ipc *ipc, struct sof_ipc_comp *comp)
 
 	platform_shared_commit(icd, sizeof(*icd));
 
-	return ret;
+	return 0;
```
Comment (lgirdwood) [inline src/ipc/ipc.c]: in fact last two patches and this could be in a separate PR for fixes as they are dont have to block on WIP status.

File: src/schedule/edf_schedule.c (hunk 3 above)
Comment (lgirdwood) [reply src/schedule/edf_schedule.c]: Lets bring this into a new PR as it may get bogged down here with other patches.

File: tools/topology/sof/tokens.m4 [hunk 8]
```diff
@@ -58,6 +58,7 @@ SectionVendorTokens."sof_comp_tokens" {
 	SOF_TKN_COMP_FORMAT			"402"
 # Token retired with ABI 3.2, do not use for new capabilities
 #	SOF_TKN_COMP_PRELOAD_COUNT		"403"
+	SOF_TKN_COMP_CORE_ID			"404"
```
[the remark this answers is not in the document]
Comment (lyakh) [reply tools/topology/sof/tokens.m4]: yes, the component runs on core X. It looks like we don't really need the default: components are typically created by `PIPELINE_PCM_ADD()` and `DAI_ADD()` macros and they already have a `core` argument. So, we just have to modify those macros to propagate that argument to all components, we don't even need to change actual platform topology files. So I would say, that it should really be set for all components. Do we still need a default?

File: src/schedule/edf_schedule.c (hunk 3 above)
Comment (lyakh) [reply src/schedule/edf_schedule.c]: @tlauda no, don't think it's because the master core doesn't wait.It doesn't matter what the master core does. The slave core while booting gets an IDC interrupt and schedules a task, before completing its own initialisation, where in the *end* of that initialisation its idle task is started.

Comment (tlauda) [reply src/schedule/edf_schedule.c]: @lyakh The only source of the IDC is master core. During my testing I've never seen this kind of race scenario, but I guess python is too slow to send another IPC before slave core boots.

Comment (lyakh) [reply src/schedule/edf_schedule.c]: @tlauda but it doesn't seem to be a real IDC, because with this fix I'm ignoring it and it isn't causing any problem

**Comment by lyakh** [comment]
fixes now moved to #2552

File: src/schedule/edf_schedule.c (hunk 3 above)
Comment (tlauda) [reply src/schedule/edf_schedule.c]: @lyakh The task is still queued, just executed later.

File: tools/topology/sof/tokens.m4 (hunk 8 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply tools/topology/sof/tokens.m4]: I would say if it's not defined by token, then it's core 0. Kernel would do this automatically by creating the IPC structure, calling bzero on structure and the parsing tokens.

File: src/schedule/edf_schedule.c (hunk 3 above)
Comment (lyakh) [reply src/schedule/edf_schedule.c]: @tlauda I did in fact try that: I added a 50ms delay in the kernel after powering up the core before sending the IPC to inform the firmware about that. Didn't help.

Comment (tlauda) [reply src/schedule/edf_schedule.c]: @lyakh Weird and very suspicious. So we are getting some kind of random IDC interrupt and don't even know what is the source...

File: tools/topology/sof/tokens.m4 (hunk 8 above)
[the remark this answers is not in the document]
Comment (lyakh) [reply tools/topology/sof/tokens.m4]: @lgirdwood is it safe to assume, that core 0 is the DSP master core? In the firmware we have a `PLATFORM_MASTER_CORE_ID` macro for that, albeit it's currently 0 on all platforms. It's probably already assumed implicitly in a couple of places in the kernel. Would be good to at least make it explicit. Anyway, yes, I'll make sure the kernel sends 0 as the core ID if none is set explicitly and I'll try to make it a bit more explicit.

File: tools/topology/m4/pga.m4 (hunk 4 above)
Comment (lgirdwood) [reply tools/topology/m4/pga.m4]: @lyakh any thoughts on this ? We should probably make it a feature.

Comment (lyakh) [reply tools/topology/m4/pga.m4]: @lgirdwood we have several levels of m4 files:
1. those, defining specific primitives in `common`, `m4`, and `platform`
2. those, defining pipelines in `sof`, using primitives from (1)
3. those, defining specific topologies in the root `topology` directory, using (2) and some of (1)

The "end user," anybody, defining or modifying a machine topology works with level (3), they hardly are expected to touch levels (1) or (2). Macros like `W_PGA()` are mostly used in level (2), they're hardly used directly in level (3). Do you propose to rework all levels or only those, used on level (3)? That would seem the most important simplification to me to make end topologies easier to work with? Also it would be good to be able to re-use definitions. In your proposal you have to define parameters for each `name`. What if we define several pipelines in a level (3) topology and some of them share the same values of parameters, like formats and sampling rate. It would be good to set such parameters once for all pipelines until they are redefined? E.g.
```
W_FORMAT(PIPELINE, s32le)
W_...
PIPELINE_PCM_ADD(1, 0)
PIPELINE_PCM_ADD(2, 1)
```
As for level (2) and (1) macros - we can simplify them too, but since they aren't exposed to end-users, maybe it's less important?

Comment (lgirdwood) [reply tools/topology/m4/pga.m4]: Good idea, the three levels make a lot of sense and we should capture this in the docs. We should also make it compulsory that  new components provide levels 1 & 2 for that component.
Yes, good idea on the name where the pipelines are the same (i.e. HDMI). Lets do level 3 first.

**Comment by lyakh** [comment]
@lgirdwood @tlauda what do we do with this one? UT failures are also present in other PRs. Do we want to merge this or is there a danger with the new token or should I first add the "core" parameter to all widget?

File: src/schedule/edf_schedule.c (hunk 3 above)
Comment (lyakh) [reply src/schedule/edf_schedule.c]: @tlauda ok, maybe that was wrong. I double-checked with the delay in the kernel between booting the slave core and sending the IPC, and the result is the same - it doesn't help. But this time I also printed in the firmware. It looks like the slave core doesn't actually boot until the IPC. Or at least it's hanging somewhere early after the power on but before initialising the EDF scheduler and only after the IDC interrupt it continues? Any idea?

Comment (tlauda) [reply src/schedule/edf_schedule.c]: @lyakh What is the exact flow from the kernel perspective. You are powering up core, unstalling and unresetting and then send power up IPC? And after that IPC if you wait a little bit the slave core still hasn't booted?

Comment (lyakh) [reply src/schedule/edf_schedule.c]: @tlauda as far as I can see, a slave core is enabled (SPA is set), then reset is cleared, then the core is unstalled. Then I insert a delay after which the kernel sends an IPC to the master core, but the slave core only boots after this last IPC.

Comment (tlauda) [reply src/schedule/edf_schedule.c]: @lyakh You need to wait after first IPC and before the next one. First IPC is to actually boot slave core out of ROM and jump to the FW.

Comment (lyakh) [reply src/schedule/edf_schedule.c]: @tlauda but then we have exactly the problem: the first IPC sends an IDC to the booting core, which then arrives before the idle task has been scheduled

Comment (tlauda) [reply src/schedule/edf_schedule.c]: @lyakh But core isn't actually booting until you send SOF_IPC_PM_CORE_ENABLE. It is stuck in ROM in endless loop. After sending this particular IPC it boots up and initialized everything.

Comment (lyakh) [reply src/schedule/edf_schedule.c]: @tlauda but it's already the first IPC that sends the first IDC: see `arch_cpu_enable_core()`.
```
	pm_runtime_get(PM_RUNTIME_DSP, id);
	pm_runtime_get(CORE_MEMORY_POW, id);
	cpu_alloc_core_context(id);
	idc_enable_interrupts(id, cpu_get_id());
	idc_send_msg(&power_up, IDC_NON_BLOCKING);
	active_cores_mask |= (1 << id);
```
So, do we take my fix or do we remove sending an IDC message from here or do we process this first IDC on the slave core without scheduling a task (effectively my approach)?

Comment (tlauda) [reply src/schedule/edf_schedule.c]: @lyakh I don't get your point. First IPC sends first IDC, because it needs to boot up the slave core. Without it you won't even get to EDF scheduler initialization. The problem is the second IDC, which is for some reason sent after this one, but before core boots up.

Comment (lyakh) [reply src/schedule/edf_schedule.c]: @tlauda ok. I thought it was the first IDC, that was already triggering the scheduler task. But that's even better then, then we really don't need to schedule in that first IDC interrupt. Maybe a better fix than special-casing in the scheduler would be not scheduling in the IDC handler if it's the POWER_UP message? That seems to be a better solution to me than waiting for some undefined time in the kernel without a way to (asynchronously) verify the core status?

Comment (tlauda) [reply src/schedule/edf_schedule.c]: @lyakh Could you check with #2593 whether issue is gone?

**Comment by lgirdwood** [comment]
Lets re-check CI, there were some CI issues lately unrelated to FW.

**Comment by lgirdwood** [comment]
SOFCI TEST

File: tools/topology/sof/tokens.m4 (hunk 8 above)
[the remark this answers is not in the document]
Comment (lyakh) [reply tools/topology/sof/tokens.m4]: @lgirdwood ping, anything blocking this one? Further, this constitutes a change of the API, doesn't it?

[the remark this answers is not in the document]
Comment (lgirdwood) [reply tools/topology/sof/tokens.m4:61]: Yep, ABI update. Will need PR for this token on kernel. You still ahve time to add to MINOR 15.

**Comment by lyakh** [comment]
What is that failure with "pr-tools-build?" It fails to build with
```
failed to compile context /home/sof/work/sof.git/tools/build_tools/test/topology/test-playback-ssp5-mclk-0-I2S-eq-fir-s32le-s32le-48k-24576k-codec.conf
ALSA lib dapm.c:307:(tplg_build_routes) warning: Undefined source widget/stream 'BUF2.1'
ALSA lib dapm.c:284:(tplg_build_routes) warning: undefined sink widget/stream 'BUF2.1'
ALSA lib dapm.c:307:(tplg_build_routes) warning: Undefined source widget/stream 'BUF2.1'
ALSA lib dapm.c:284:(tplg_build_routes) warning: undefined sink widget/stream 'BUF2.1'
ALSA lib dapm.c:307:(tplg_build_routes) warning: Undefined source widget/stream 'BUF2.1'
ALSA lib dapm.c:284:(tplg_build_routes) warning: undefined sink widget/stream 'BUF2.1'
```
And I don't even know how to build that directory. Ideas?

File: tools/topology/sof/tokens.m4 (hunk 8 above)
[the remark this answers is not in the document]
Comment (lyakh) [reply tools/topology/sof/tokens.m4:61]: @lgirdwood ok, now comes the confusion. How do we do that? I saw multiple PRs with ABI updates, each trying to modify the abi.h header in their own incompatible way. What's the correct way? Just do the change (add the token) and not touch abi.h at all? To have it modified later by a separate commit?

**Comment by lgirdwood** [comment]
@lyakh can you link the kernel PR to this PR, sorry I cant see it.

**Comment by lyakh** [comment]
Something broke down. Looks like FIR EQ expects some fixed structure? Adding a token to it broke it:
```
[   19.124250] [1184] sof-audio-pci 0000:00:1f.3: ASoC: creating DAPM widget EQIIR2.0 id 18
[   19.124258] [1184] sof-audio-pci 0000:00:1f.3: ASoC: template EQIIR2.0 has 1 controls of type 2
[   19.124263] [1184] sof-audio-pci 0000:00:1f.3: ASoC: adding bytes kcontrol eqiir_bytes_2 with access 0x10000030
[   19.124269] [1184] sof-audio-pci 0000:00:1f.3: tplg: load control type 2 name : eqiir_bytes_2
[   19.124275] [1184] sof-audio-pci 0000:00:1f.3: tplg: load kcontrol index 11 chans 0
[   19.124280] sof-audio-pci 0000:00:1f.3: error: Wrong ABI magic 0x00000014.
[   19.124291] sof-audio-pci 0000:00:1f.3: ASoC: failed to init eqiir_bytes_2
[   19.124298] sof-audio-pci 0000:00:1f.3: ASoC: failed to load widget EQIIR2.0
```

**Comment by lyakh** [comment]
> @lyakh can you link the kernel PR to this PR, sorry I cant see it.

@lgirdwood you cannot see it because it isn't there yet :-) As I wrote yesterday this version of the PR is breaking things. So, I first have to fix it before ABI bumping would make sense, right? I'll see if I can hunt it down today.

**Comment by lyakh** [comment]
> Something broke down. Looks like FIR EQ expects some fixed structure? Adding a token to it broke it:
> 
> ```
> [   19.124250] [1184] sof-audio-pci 0000:00:1f.3: ASoC: creating DAPM widget EQIIR2.0 id 18
> [   19.124258] [1184] sof-audio-pci 0000:00:1f.3: ASoC: template EQIIR2.0 has 1 controls of type 2
> [   19.124263] [1184] sof-audio-pci 0000:00:1f.3: ASoC: adding bytes kcontrol eqiir_bytes_2 with access 0x10000030
> [   19.124269] [1184] sof-audio-pci 0000:00:1f.3: tplg: load control type 2 name : eqiir_bytes_2
> [   19.124275] [1184] sof-audio-pci 0000:00:1f.3: tplg: load kcontrol index 11 chans 0
> [   19.124280] sof-audio-pci 0000:00:1f.3: error: Wrong ABI magic 0x00000014.
> [   19.124291] sof-audio-pci 0000:00:1f.3: ASoC: failed to init eqiir_bytes_2
> [   19.124298] sof-audio-pci 0000:00:1f.3: ASoC: failed to load widget EQIIR2.0
> ```

I haven't traced it down to the exact code line, but everything is pointing at binary equaliser coefficients. I'm comparing topology files in their binary presentation and I see that the coefficient block is moved down in the patched version and then the kernel driver doesn't find the magic number. @singalsu @juimonen any ideas?

File: tools/topology/m4/bytecontrol.m4 [hunk 9]
```diff
@@ -55,7 +66,8 @@ define(`C_CONTROLBYTES',
 `		tlv_callback'
 `	]'
 `	data ['
-`		$10'
+`		"'N_CONTROLBYTES($2)`_data_w"'
+`		$11'
```
[the remark this answers is not in the document]
Comment (lyakh) [reply tools/topology/m4/bytecontrol.m4]: thanks,  @juimonen I think you're right - I don't need to modify `C_CONTROLMIXER()` and `C_CONTROLBYTES()` at all! Patch set updated without those two patches.

File: tools/topology/sof/tokens.m4 (hunk 8 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply tools/topology/sof/tokens.m4:61]: yes you are ok for 15 without making a abi.h change today.

[the remark this answers is not in the document]
Comment (lyakh) [reply tools/topology/sof/tokens.m4:61]: @lgirdwood https://github.com/thesofproject/linux/pull/1981 All CI passed except for one "still running" since last Friday.

[the remark this answers is not in the document]
Comment (kv2019i) [reply tools/topology/sof/tokens.m4:61]: @lgirdwood FYI, we hit a consensus gap w.r.t. what is actually required for the driver side of an ABI change -> https://github.com/thesofproject/linux/pull/1981#issuecomment-610343522

**Comment by lgirdwood** [comment]
CI looks good, so far, but jenkins was down. @lyakh can you fix conflicts, this will also reschedule on jenkins too so we should be able to merge when complete.

**Comment by lyakh** [comment]
@tlauda @lrgirdwo if we want to merge this series, would be cool to do it before another 27 pipeline topologies are merged (no, @fredoh9 I'm not looking at you)

**Comment by lgirdwood** [comment]
@lyakh yep, it's next in line :-)

**Comment by lgirdwood** [comment]
@zrombel can you comment on CI, it seems blocked ?

**Comment by lgirdwood** [comment]
@zrombel forget it, it's just magically completed :-)

**Comment by lgirdwood** [comment]
Jenkins known issues.
